### PR TITLE
Added node16 execution handler

### PR DESCRIPTION
**Pull request by AndreyIvanov42**
Added node16 execution handler - to be able to execute tasks using this version.
Mostly similar to https://github.com/microsoft/azure-pipelines-agent/pull/3103

# Testing
- Changes have been tested on Linux, MacOS, Windows: 
  by unit-tests
  manually as part of some task. Used tasks and cases are described [here](https://github.com/microsoft/azure-pipelines-agent/pull/3861#issuecomment-1169892560)
- Enough test coverage for changes and current test coverage for the task doesn't look poor
- I didn't find any breaking changes
- There are no other concerns
- I haven't discovered any new uncovered test/use cases

### CODE REVIEW

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 1]
```diff
@@ -171,10 +171,16 @@ public async Task RunAsync()
...
             bool useNode10 = AgentKnobs.UseNode10.GetValue(ExecutionContext).AsBoolean();
+            bool useNode16 = AgentKnobs.UseNode16.GetValue(ExecutionContext).AsBoolean();
             bool taskHasNode10Data = Data is Node10HandlerData;
+            bool taskHasNode16Data = Data is Node16HandlerData;
 
             string nodeFolder = "node";
-            if (taskHasNode10Data || useNode10)
+            if (taskHasNode16Data || useNode16)
+            {
+                Trace.Info($"Task.json has node16 handler data: {taskHasNode16Data}");
```
Comment (alexander-smolyakov) [inline src/Agent.Worker/Handlers/NodeHandler.cs:181]:
```suggestion
                Trace.Info($"Task.json has node16 handler data: {taskHasNode16Data}, use node16 for node tasks: {useNode16}");
```

Let's show the status of the knob as well.

File: src/Misc/externals.sh [hunk 2]
```diff
@@ -173,6 +176,8 @@ if [[ "$PACKAGERUNTIME" == "win-x86" ]]; then
     fi
     acquireExternalTool "$NODE_URL/v${NODE10_VERSION}/win-x86/node.exe" node10/bin
     acquireExternalTool "$NODE_URL/v${NODE10_VERSION}/win-x86/node.lib" node10/bin
+    acquireExternalTool "$NODE_URL/v${NODE16_VERSION}/win-x64/node.exe" node16/bin
+    acquireExternalTool "$NODE_URL/v${NODE16_VERSION}/win-x64/node.lib" node16/bin
```
Comment (alexander-smolyakov) [inline src/Misc/externals.sh:180]: I believe there should be `win-x86` architecture

```suggestion
    acquireExternalTool "$NODE_URL/v${NODE16_VERSION}/win-x86/node.exe" node16/bin
    acquireExternalTool "$NODE_URL/v${NODE16_VERSION}/win-x86/node.lib" node16/bin
```

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 3]
```diff
@@ -171,10 +171,16 @@ public async Task RunAsync()
...
+                nodeFolder = "node16";
+            } else if (taskHasNode10Data || useNode10)
```
Comment (alexander-smolyakov) [inline src/Agent.Worker/Handlers/NodeHandler.cs:183]: Please check the code formatting

```suggestion
            } 
            else if (taskHasNode10Data || useNode10)
```

**Review by alexander-smolyakov — COMMENTED** [review]
Please take a look at the comments

**Comment by alexander-smolyakov** [comment]
Could you please update the PR description as well?

**Comment by AndreyIvanov42** [comment]
> Could you please update the PR description as well?

Done

**Comment by AndreyIvanov42** [comment]
## Migrated task 
[CopyFiles](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/copy-files?view=azure-devops&tabs=yaml), [UsePythonVersion](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/use-python-version?view=azure-devops), [CMake](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/cmake?view=azure-devops)

## Environment used (OS/Agent)
Windows 10x64 / Agent with Node 16 handler, Ubuntu 16.04x64 / Agent with Node 16 handler, macOS 12 / Agent with Node 16 handler

## Test cases

```
- task: CMake@1
  inputs:
    workingDirectory: 'src'
    cmakeArgs: -DCMAKE_SHELL=false -DTEST_ONE=one;two -DTEST_TWO=/one/two/three -DTEST_WIN_VARIABLE=test -DTEST_VARIABLE=test
    runInsideShell: false 
```

```
- task: CMake@1
  inputs:
    workingDirectory: 'src'
    cmakeArgs: -DCMAKE_SHELL=true -DTEST_ONE=one;two -DTEST_TWO=/one/two/three -DTEST_VARIABLE=$USER -DTEST_WIN_VARIABLE=%OS%
    runInsideShell: true
``` 
  
```
- task: UsePythonVersion@0
  inputs:
    versionSpec: 3.10.5
    disableDownloadFromRegistry: true
```

```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents: *.js
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: false
    OverWrite: false
    flattenFolders: false
    preserveTimestamp: true
```

```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents: *.js
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: false
    OverWrite: true
    flattenFolders: false
    preserveTimestamp: true
```

```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents: *
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: true
    OverWrite: false
    flattenFolders: false
    preserveTimestamp: false
```
  
```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents: **/bin/**
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: true
    OverWrite: false
    flattenFolders: true
    preserveTimestamp: false
```

```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents: **
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: true
    OverWrite: false
    flattenFolders: true
    preserveTimestamp: false
```

```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents: |
      *.css
      **/bin/**.css
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: true
    OverWrite: false
    flattenFolders: true
    preserveTimestamp: false
```

```
- task: CopyFiles@2
  inputs:
    SourceFolder: 'src/copy-files/source'
    Contents:  root-file1.js
    TargetFolder: 'src/copy-files/target'
    CleanTargetFolder: true
    OverWrite: false
    flattenFolders: false
    preserveTimestamp: true
```

**Review by DenisRumyantsev — APPROVED** [review]

**Review by kirill-ivlev — APPROVED** [review]

File: src/Misc/externals.sh [hunk 4]
```diff
@@ -10,6 +10,7 @@ CONTAINER_URL=https://vstsagenttools.blob.core.windows.net/tools
 NODE_URL=https://nodejs.org/dist
 NODE_VERSION="6.17.1"
 NODE10_VERSION="10.24.1"
+NODE16_VERSION="16.13.0"
```
Comment (JamieMagee) [inline src/Misc/externals.sh:13]: Latest version
```suggestion
NODE16_VERSION="16.15.1"
```

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 5]
```diff
@@ -172,11 +172,22 @@ public string GetNodeLocation()
...
+            else if (taskHasNode10Data)
+            {
+                Trace.Info($"Task.json has node10 handler data: {taskHasNode10Data}");
+                nodeFolder = "node10";
+            }
+            else if (useNode10)
+            {
+                Trace.Info($"Found UseNode10 knob, use node10 for node tasks: {useNode10}");
                 nodeFolder = "node10";
             }
```
Comment (alexander-smolyakov) [inline src/Agent.Worker/Handlers/NodeHandler.cs:192]: Nit: Could it make sense to move this logic to a separate private method and call it something like `ResolveNodeVersion`?

**Review by alexander-smolyakov — APPROVED** [review]
LGTM and please take a look at the comment below

File: src/Agent.Worker/Handlers/NodeHandler.cs (hunk 5 above)
Comment (AndreyIvanov42) [reply src/Agent.Worker/Handlers/NodeHandler.cs:192]: I tried moving it to a separate function, but only the `Path.combine` logic remains in the original function in this case

**Review by wojgasior — APPROVED** [review]

**Pull request merged**